Even Pods Spread - 5. Priority Core

Huang-Wei opened the pull request:
**What type of PR is this?**

/sig scheduling
/kind feature
/priority important-soon
/hold

/assign @bsalamat 
/cc @krmayankk 

**What this PR does / why we need it**:

This is the **5th** PR of the "Even Pods Spread" KEP implementation. After this PR, users can run workloads using "hard/soft topologySpreadConstraints", and high priority Pod can preempt low priority one.

- Define a new internal Priority to supports API spec `whenUnsatisfiable: ScheduleAnyway`
- Core priority logic
- Unit tests

**Which issue(s) this PR fixes**:

Part of #77284.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

(will document all the changes in one place)

```release-note
NONE
```

Huang-Wei committed:
EvenPodsSpread: Define a new Priority

Huang-Wei committed:
EvenPodsSpread: Core Priority logic

Huang-Wei committed:
EvenPodsSpread: Make some funcs in predicates pkg as public

Huang-Wei committed:
EvenPodsSpread: minor enhancement on printing out priority score

Huang-Wei committed:
[eps-priority] auto-gen files

Huang-Wei commented:
some discussion on the Normalize function: https://kubernetes.slack.com/archives/C09TP78DV/p1557473288102500

ahg-g commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,216 @@
+		// need to reverse b/c the more matching pods it has, the less qualified it is
+		// result[i].Score = schedulerapi.MaxPriority - int(fScore)
+		result[i].Score = int(fScore)
+	}
```
this we can run in parallel as well.

alculquicondor replied:
It seems like a not very expensive loop (we only iterate through nodes, not pods). It might require perf-testing.

ahg-g replied:
Probably, it depends on scale and cluster state though. Not worth doing it now though, I agree.

ahg-g reviewed (commented):
similar to the previous PRs, this looks good to me as an initial implementation.

alculquicondor commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -15,3 +15,196 @@ limitations under the License.
+// that are with WhenUnsatifiable=ScheduleAnyway (a.k.a soft constraint).
+// For each node (not only "filtered" nodes by Predicates), it adds the number of matching pods
+// (all topologySpreadConstraints must be satified) as a "weight" to any "filtered" node
+// which has the <topologyKey:value> pair present.
```
I think this comment could start with a more high level (and more readable) sentence. What we are trying to say is that, for each node, we want to sum up the number of pods in all the nodes in its same topology domain. But also mention that we reverse that number? (total-counts[node])

alculquicondor commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -15,3 +15,196 @@ limitations under the License.
+		t.counts[node.Name] = &count
+	}
+	// calculate final priority score for each node
+	// TODO(Huang-Wei): in alpha version, we keep the formula as simple as possible.
```
Consider creating an issue for this

alculquicondor replied:
In my mind, if the node satisfy the constraints as if they were hard, it's priority should be `schedulerapi.MaxPriority`.

Huang-Wei replied:
In that case, we have to check the formula which is implemented in Predicates again, that calculation effort isn't trivial. IMO in terms of Priority, it's not worth that way. The current algorithm also can rank the nodes well.

alculquicondor replied:
We should be able to reuse calculations once we move to the framework. Can you open a tracking bug to consider a new formula?

ahg-g replied:
we can't reuse the calculations because the predicate metadata is only computed for the hard constraints, not all constraints.

Huang-Wei replied:
@ahg-g is right, they are separate constraints :)

alculquicondor replied:
I know... but the calculation is the same, if I'm not missing something. We should be able to iterate only once through all the nodes for all the constraints.

ahg-g replied:
I don't think the calculation is the same, we don't iterate over the soft constraints now at all, and so they are not matched against the node or the pods on the nodes. If we do that, and the node got filtered out before getting scored, that calculation becomes a waste.

alculquicondor commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -15,3 +15,196 @@ limitations under the License.
+			continue
+		}
+		fScore := float64(schedulerapi.MaxPriority) * (float64(*t.counts[node.Name]-minCount) / float64(maxMinDiff))
+		// need to reverse b/c the more matching pods it has, the less qualified it is
```
Didn't we reverse in the loop above? Otherwise, is this a TODO, or maybe just a leftover?

Huang-Wei replied:
Sry, a leftover here...

draveness commented on pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -87,10 +87,15 @@ func ApplyFeatureGates() {
 		factory.RegisterFitPredicate(predicates.EvenPodsSpreadPred, predicates.EvenPodsSpreadPredicate)
+		// register priority
+		factory.InsertPriorityKeyToAlgorithmProviderMap(priorities.EvenPodsSpreadPriority)
+		factory.RegisterPriorityFunction(priorities.EvenPodsSpreadPriority, priorities.CalculateEvenPodsSpreadPriority, 1)
```
This function has already been deprecated, we should use map-reduce pattern instead of this. I created https://github.com/kubernetes/kubernetes/pull/80225 to remove this from the codebase.

> // RegisterPriorityFunction registers a priority function with the algorithm registry. Returns the name,
// with which the function was registered.
// DEPRECATED
// Use Map-Reduce pattern for priority functions.

draveness replied:
Friendly reminder on the usage of deprecated `RegisterPriorityFunction`. :)

Huang-Wei replied:
#80225 needs further considerations. I will take another round of review later.

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,216 @@
+
+// CalculateEvenPodsSpreadPriority computes a score by checking through the topologySpreadConstraints
+// that are with WhenUnsatifiable=ScheduleAnyway (a.k.a soft constraint).
+// For each node (not only "filtered" nodes by Predicates), it adds the number of matching pods
```
SG - yes, I meant to say "all nodes" (not only the nodes that passed the predicates)

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,216 @@
+// which has the <topologyKey:value> pair present.
+// Then the sumed "weight" are normalized to 0~10, and the node(s) with the highest score are
+// the most preferred.
+// Symmetry is not considered.
```
It means we only weigh how incomingPod matches existingPod. And how existingPod matches incomingPod doesn't contribute to the final score. This is different with Affinity API.

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,216 @@
+
+func (t *topologySpreadConstrantsMap) initialize(pod *v1.Pod, nodes []*v1.Node) {
+	constraints := getSoftTopologySpreadConstraints(pod)
+	for _, node := range nodes {
```
We can.

Huang-Wei replied:
Maybe not worth it due to the cost of lock. And here `nodes` are the filtered nodes.

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,216 @@
+type topologySpreadConstrantsMap struct {
+	// The first error that we faced.
+	firstError error
+	sync.Mutex
```
Updated.

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,216 @@
+			for _, pair := range pairs {
+				t.topologyPairToNodeNames[pair] = append(t.topologyPairToNodeNames[pair], node.Name)
+			}
+			t.counts[node.Name] = new(int64)
```
A little background on using *int64 over int64 is b/c with *int64, we can leverage atomic.AddInt64 to eliminate lock. The cost of allocating int64 pointers (esp. it's an on demand initialization) outweighs the cost of locking/unlocking.

Huang-Wei replied:
Nvm, in the latest version I almost rewrite the Priority function and adopted this suggestion.

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -15,3 +15,196 @@ limitations under the License.
+	"k8s.io/kubernetes/pkg/scheduler/algorithm/predicates"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+
```
It seems a convention to keep "k8s.io/klog" apart with other imports.

Huang-Wei committed:
EvenPodsSpread: weigh constraints individually

- update logic to weigh each constraint individually
- address comments and misc fixes

alculquicondor commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,210 @@
+		} else if count < minCount {
+			minCount = count
+		}
+		t.podCounts[node.Name] = &count
```
If all nodes end up with the same count, they seem to get 0 priority. Is that ok?

Huang-Wei replied:
Nope. They will be all get 10. I added a testcase covering this.

alculquicondor commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,210 @@
+			// And we add matchCount to <t.total> to reverse the final score later.
+			for _, nodeName := range t.topologyPairToNodeNames[pair] {
+				atomic.AddInt64(t.podCounts[nodeName], matchCount[tpKey])
+				atomic.AddInt64(&t.total, matchCount[tpKey])
```
At the end of the day, total is a constant for all nodes (we normalize between min and max score). So it's meaningless when calculating the priority.

Huang-Wei replied:
Below is how I think about this problem:

If we don't have a `total` value, we only know the number of matching pods on each node, that's not enough. As the number of matching pods sort of represents the credit it "owe", but without `total`, we don't have a global picture how many credits the whole system "owe"s, hence can't know the "owed" credit ratio for each node.

For example, two nodes are matched with 5/6 pods, some other priorities use an algorithm like this: find the max(6), min defaults to 0, then the first node gets a score 2 (= 10-50/6), and the second gets 0!! (= 10-60/6) That's not a good normalization.

With `total`, we can give better math: reverse the matching to be 6/5 (11 -5or6), then give the final score: the first node gets 10 (= 60/6), and the second gets 8 (= 50/6).

alculquicondor replied:
Right, if you normalize between best and worst node, then it doesn't serve any purpose. But you are (now?) normalizing between 0 and best node.

Huang-Wei replied:
Yes, the old commit had `minCount`, but that also ends up with 0. So given that matching number can't be negative, I assumed `minCount` to be 0.

alculquicondor commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,210 @@
+		}
+		if match {
+			for _, pair := range pairs {
+				t.topologyPairToNodeNames[pair] = append(t.topologyPairToNodeNames[pair], node.Name)
```
You can avoid these lists doing the following:
For each topology pair, calculate how many pods match `O(P*C)`. Then iterate through the nodes and sum up the counters for each of its topologies `O(N*C)`. Total `O(P*C)`.

P=# pods, C=# constraints, N=# nodes.

At first I thought your algorithm was `O(N^2*C)`, because you iterate through nodes within your parallel for. But it actually amortizes to `O(P*C)` plus the same `O(N*C)`, because a node only belongs to one set of topologies. In any case, you can get rid of the per-node map with a per-topologypair map. And sum up later for each node.

This can all be done in a single parallel for on the nodes, if you use appropriate synchronization primitives (like wait group).

Huang-Wei replied:
Per our offline talk, we should note that `N` can be `N1` (all nodes) and `N2` (filtered nodes which passed the Predicates). So generally speaking, no algorithm can do better than `N1` * `P` (P = pods in **one** node) * `C`. I put an additional `N2` loop in `initilize` is to filter out non-qualified nodes, as well as do some prepping data structures.

Please check out the latest version. It also includes some benchmark tests. It's doing a much better job.

```
## Old Commit

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     300	   3880228 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     500	   4121098 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     500	   4084106 ns/op

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	     500	   3704594 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	     300	   5029568 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	     500	   3408629 ns/op

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     200	   7106261 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     200	   7428552 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     200	   6032280 ns/op

## New Commit

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     500	   3133159 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     500	   2814902 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     500	   3206365 ns/op

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	    1000	   1802340 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	    1000	   1869673 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	    1000	   1927666 ns/op

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     300	   4197458 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     300	   4181693 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     300	   4242142 ns/op
```

alculquicondor replied:
Btw, this looks awesome.

Huang-Wei committed:
EvenPodsSpread: optimize Priority logic

Huang-Wei committed:
EvenPodsSpread: Benchmarking Priority function

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -15,3 +15,196 @@ limitations under the License.
+			}
+			matchCount := 0
+			for _, existingPod := range nodeInfo.Pods() {
+				match, err := predicates.PodMatchesAllSpreadConstraints(existingPod, pod.Namespace, constraints)
```
Updated to respect the individual constraint independently.

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,210 @@
+	}
+}
+
+// CalculateEvenPodsSpreadPriority computes a score by checking through the topologySpreadConstraints
```
I haven't updated it yet. Will revisit tomorrow.

draveness commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,220 @@
+// Note: Symmetry is not applicable. We only weigh how incomingPod matches existingPod.
+// Whether existingPod matches incomingPod doesn't contribute to the final score.
+// This is different with the Affinity API.
+func CalculateEvenPodsSpreadPriority(pod *v1.Pod, nodeNameToInfo map[string]*schedulernodeinfo.NodeInfo, nodes []*v1.Node) (schedulerapi.HostPriorityList, error) {
```
Where do we call this priority function? This is the only place "CalculateEvenPodsSpreadPriority" string appears.

Huang-Wei replied:
Good question.. I realized the issue only until I wrote the integration test (i.e. in PR6, you can see the fix - it's actually an typo in `defaults.go`...)

But as you raised this, to avoid confusion, let me move that fix in PR6 to this PR.

BTW: we can see the benefits of having different kinds of tests :)

Huang-Wei replied:
Oh... In addition to that typo, I made another rebasing mistake..

Huang-Wei replied:
Done. Squashed to commit "EvenPodsSpread: Define a new Priority".

Thanks for the catch! (Otherwise, the next PR is the only guard...)

To get the benefits of making each PR CI green, I paid way too much energy on rebasing...

draveness replied:
> Done. Squashed to commit "EvenPodsSpread: Define a new Priority".
> 
> Thanks for the catch! (Otherwise, the next PR is the only guard...)
> 
> To get the benefits of making each PR CI green, I paid way too much energy on rebasing...

The rebasing could cause tons of work. Thanks for breaking this into pieces.

alculquicondor commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,220 @@
+		for _, constraint := range constraints {
+			if tpVal, ok := node.Labels[constraint.TopologyKey]; ok {
+				pair := topologyPair{key: constraint.TopologyKey, value: tpVal}
+				matchSum := atomic.LoadInt64(t.topologyPairToPodCounts[pair])
```
This is not running in parallel, so no need to use atomic. Same below.

Huang-Wei replied:
This case is a _little_ tricky b/c we need to **concurrently** compare with `minCount`, and it's a racing behavior. We _may_ need to use `atomic.CompareAndSwapInt64`. Will give a try on this.

Huang-Wei replied:
I did a test ([gist](https://gist.github.com/Huang-Wei/7fa9eafb7074e3b4556211aa4723afe4)) and seems parallel version doesn't help much.

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,220 @@
+	// nodeNameToPodCounts is keyed with node name, and valued with the number of matching pods.
+	nodeNameToPodCounts map[string]int64
+	// topologyPairToPodCounts is keyed with topologyPair, and valued with the number of matching pods.
+	topologyPairToPodCounts map[topologyPair]*int64
```
Unfortunately no. Golang doesn't support [addressing a map value](https://github.com/golang/go/issues/11865), i.e. if `topologyPairToPodCounts` is defined as `map[topologyPair]int64`, we can't operate `&topologyPairToPodCounts[pair]`.

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,220 @@
+		} else {
+			// For those nodes which don't have all required topologyKeys present, it's intentional to
+			// assign nodeNameToPodCounts[nodeName] as -1, so that we're able to score these nodes to 0 afterwards.
+			t.nodeNameToPodCounts[node.Name] = -1
```
Good point. Will update.

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,220 @@
+
+		for _, constraint := range constraints {
+			pair := topologyPair{key: constraint.TopologyKey, value: node.Labels[constraint.TopologyKey]}
+			// If current topology pair is not associated with any "filtered" node,
```
"candidate node" sounds good.

Huang-Wei commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,220 @@
+// that are with WhenUnsatisfiable=ScheduleAnyway (a.k.a soft constraint).
+// The function works as below:
+// 1) In all nodes, calculate the number of pods which match <pod>'s soft topology spread constraints.
+// 2) Sum up the number to each node in <nodes> which has corresponding topologyPair present.
```
> did you mean "sum up the number of pods in each node in "?

yes, I will rephrase to "sum up the number calculated in 1) to ..."

> perhaps rephrase the last part to "has all topologyKeys and matches affinity terms of the incoming pod"?

Not exactly. It's been rephrased another way, please see if it makes sense.

Huang-Wei commented:
Comments addressed. @ahg-g @alculquicondor PTAL.

ahg-g commented:
/lgtm

Congrats on landing this feature!

Huang-Wei commented:
Phew, one PR left. @ahg-g thanks for your reviews as usual.

BTW: I squashed and reordered the commits slightly. Will request a /lgtm later.

Huang-Wei commented:
@ahg-g @alculquicondor @draveness would you mind re-labeling /lgtm?

draveness reviewed (approved):
/lgtm

Thanks @Huang-Wei !

Huang-Wei commented:
/hold cancel

tedyu commented on pkg/scheduler/algorithm/priorities/even_pods_spread.go:
```diff
@@ -0,0 +1,209 @@
+
+	errCh := schedutil.NewErrorChannel()
+	ctx, cancel := context.WithCancel(context.Background())
+	processAllNode := func(i int) {
```
processNode seems to be better name.

Huang-Wei replied:
That's the old name. Here I want to highlight that this function works for **all** nodes, instead of each **candidate** node which has passed Predicates check.
